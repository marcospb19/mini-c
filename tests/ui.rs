//! This is a test module that creates snapshots, and might do
//! UI testing in the future.
//!
//! This shouldn't be called ui.rs hmmmmmmmm.
//!
//! Snapshot todo list:
//! - [X] Source code.
//! - [X] AST.
//! - [ ] LLVM IR.
//! - [ ] Execution stdout output (UI).
//! - [ ] (maybe) Execution final state (variables, functions).

use fs_err as fs;
use mini_c::test_utils;

macro_rules! snapshot_example_file {
    ($example_path:literal) => {
        let file_contents = &fs::read_to_string($example_path)
            .unwrap_or_else(|err| panic!("Failed to read file {}: {err}.", $example_path));
        // Source code
        insta::assert_display_snapshot!(file_contents);
        // AST
        insta::assert_ron_snapshot!(test_utils::generate_ast(file_contents));
    };
}

#[test]
fn example_break() {
    snapshot_example_file!("examples/break.c");
}
#[test]
fn example_coins() {
    snapshot_example_file!("examples/coins.c");
}
#[test]
fn example_conds() {
    snapshot_example_file!("examples/conds.c");
}
#[test]
fn example_easy() {
    snapshot_example_file!("examples/easy.c");
}
#[test]
fn example_fib() {
    snapshot_example_file!("examples/fib.c");
}
#[test]
fn example_file_1() {
    snapshot_example_file!("examples/file_1.c");
}
#[test]
fn example_file_2() {
    snapshot_example_file!("examples/file_2.c");
}
#[test]
fn example_if_pass() {
    snapshot_example_file!("examples/if-pass.c");
}
#[test]
fn example_operators() {
    snapshot_example_file!("examples/operators.c");
}
#[test]
fn example_queen() {
    snapshot_example_file!("examples/queen.c");
}
#[test]
fn example_var() {
    snapshot_example_file!("examples/var.c");
}
#[test]
fn example_var_loop() {
    snapshot_example_file!("examples/var_loop.c");
}
#[test]
fn example_while() {
    snapshot_example_file!("examples/while.c");
}
