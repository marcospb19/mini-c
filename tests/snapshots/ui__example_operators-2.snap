---
source: tests/ui.rs
expression: test_mini_c(file_contents)
---
Program(
  preamble: true,
  declarations: [
    Function(FunctionDeclaration(Int, "main", [], Some(Scope([
      VariableDeclaration(Int, [
        Normal("a"),
      ]),
    ], [
      Expression(Assignment(AssignmentExpression(Normal("a"), Value(Int(59))))),
      If(IfStatement(Binary(BinaryExpression(VariableReference(Normal("a")), Equals, Value(Int(59)))), Scope([], [
        Expression(FunctionCall(FunctionCallExpression("putint", [
          Value(Int(1)),
        ]))),
      ]), Some(Scope([], [
        Expression(FunctionCall(FunctionCallExpression("putint", [
          Unary(UnaryExpression(Negative, Value(Int(1)))),
        ]))),
      ])))),
      If(IfStatement(Binary(BinaryExpression(Unary(UnaryExpression(Negative, VariableReference(Normal("a")))), Equals, Unary(UnaryExpression(Negative, Value(Int(59)))))), Scope([], [
        Expression(FunctionCall(FunctionCallExpression("putint", [
          Value(Int(2)),
        ]))),
      ]), Some(Scope([], [
        Expression(FunctionCall(FunctionCallExpression("putint", [
          Unary(UnaryExpression(Negative, Value(Int(2)))),
        ]))),
      ])))),
      If(IfStatement(Binary(BinaryExpression(Binary(BinaryExpression(Binary(BinaryExpression(Binary(BinaryExpression(VariableReference(Normal("a")), Add, Value(Int(1)))), Add, Value(Int(2)))), Sub, Value(Int(1)))), NotEquals, Value(Int(61)))), Scope([], [
        Expression(FunctionCall(FunctionCallExpression("putint", [
          Unary(UnaryExpression(Negative, Value(Int(3)))),
        ]))),
      ]), Some(Scope([], [
        Expression(FunctionCall(FunctionCallExpression("putint", [
          Value(Int(3)),
        ]))),
      ])))),
      If(IfStatement(Unary(UnaryExpression(Not, Binary(BinaryExpression(Binary(BinaryExpression(Binary(BinaryExpression(Binary(BinaryExpression(VariableReference(Normal("a")), Mul, Value(Int(10)))), Mul, Value(Int(10)))), Div, Value(Int(20)))), Equals, Binary(BinaryExpression(Value(Int(59)), Mul, Value(Int(5)))))))), Scope([], [
        Expression(FunctionCall(FunctionCallExpression("putint", [
          Unary(UnaryExpression(Negative, Value(Int(4)))),
        ]))),
      ]), Some(Scope([], [
        Expression(FunctionCall(FunctionCallExpression("putint", [
          Value(Int(4)),
        ]))),
      ])))),
      If(IfStatement(Binary(BinaryExpression(Binary(BinaryExpression(Value(Bool(true)), And, Value(Bool(false)))), And, Value(Bool(false)))), Scope([], [
        Expression(FunctionCall(FunctionCallExpression("putint", [
          Unary(UnaryExpression(Negative, Value(Int(5)))),
        ]))),
      ]), Some(Scope([], [
        Expression(FunctionCall(FunctionCallExpression("putint", [
          Value(Int(5)),
        ]))),
      ])))),
      If(IfStatement(Binary(BinaryExpression(Binary(BinaryExpression(Value(Bool(false)), Or, Value(Bool(true)))), Or, Value(Bool(false)))), Scope([], [
        Expression(FunctionCall(FunctionCallExpression("putint", [
          Value(Int(6)),
        ]))),
      ]), Some(Scope([], [
        Expression(FunctionCall(FunctionCallExpression("putint", [
          Unary(UnaryExpression(Negative, Value(Int(6)))),
        ]))),
      ])))),
      If(IfStatement(Binary(BinaryExpression(Binary(BinaryExpression(Binary(BinaryExpression(Value(Int(9)), Less, Value(Int(10)))), And, Binary(BinaryExpression(Value(Int(9)), Greater, Value(Int(10)))))), And, Binary(BinaryExpression(Value(Int(9)), Equals, Value(Int(9)))))), Scope([], [
        Expression(FunctionCall(FunctionCallExpression("putint", [
          Value(Int(7)),
        ]))),
      ]), Some(Scope([], [
        Expression(FunctionCall(FunctionCallExpression("putint", [
          Unary(UnaryExpression(Negative, Value(Int(7)))),
        ]))),
      ])))),
      If(IfStatement(Binary(BinaryExpression(Binary(BinaryExpression(Binary(BinaryExpression(Value(Int(10)), Greater, Value(Int(9)))), And, Binary(BinaryExpression(Value(Int(10)), Less, Value(Int(9)))))), And, Binary(BinaryExpression(Value(Int(9)), Equals, Value(Int(9)))))), Scope([], [
        Expression(FunctionCall(FunctionCallExpression("putint", [
          Value(Int(8)),
        ]))),
      ]), Some(Scope([], [
        Expression(FunctionCall(FunctionCallExpression("putint", [
          Unary(UnaryExpression(Negative, Value(Int(8)))),
        ]))),
      ])))),
      If(IfStatement(Unary(UnaryExpression(Not, Binary(BinaryExpression(Value(Int(10)), NotEquals, Value(Int(10)))))), Scope([], [
        Expression(FunctionCall(FunctionCallExpression("putint", [
          Value(Int(9)),
        ]))),
      ]), Some(Scope([], [
        Expression(FunctionCall(FunctionCallExpression("putint", [
          Unary(UnaryExpression(Negative, Value(Int(9)))),
        ]))),
      ])))),
      Return(ReturnStatement(Some(Value(Int(0))))),
    ])))),
  ],
)
